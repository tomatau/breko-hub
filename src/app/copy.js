import React from 'react'

export const app = {
  loading: `Loading...`,
  title: `Breko Hub`,
  meta: {
    description: `A boilerplate for building universal react applications`,
    keywords: `react,redux,react-router,koa,universal,babel,webpack`,
  },
}

export const nav = {
  home: `Home`,
  bar: `Bar`,
  private: `Admin`,
}

export const homeRoute = {
  content: (
    <>
      <h2>Welcome to Breko-Hub!</h2>
      <p>A boilerplate designed for quick development!</p>
      <p>There are examples showing how you can customise to your needs.</p>
      <p>The server routes also auto update on changes!</p>
    </>
  ),
}

export const barRoute = {
  documentTitle: `Bar`,
  content: (
    <>
      <h2>The Bar Route!</h2>
      <p>API request to the <code>/api/bar</code> endpoint.</p>
    </>
  ),
}

export const privateRoute = {
  documentTitle: `Confidential`,
  flashMessage: `You were redirected because you can't view the private page`,
  content: (
    <>
      <h2>You shouldn&apos;t be here!</h2>
      <p>You may not view the private route!</p>
    </>
  ),
}

export const notFoundRoute = {
  documentTitle: `Doesn't exist`,
  content: (
    <>
      <h2>404</h2>
      <p>Page not found!</p>
    </>
  ),
}

export const oopsRoute = {
  documentTitle: `Oopsies`,
  content: (
    <>
      <h2>There was a problem</h2>
      <p>Something went oops, oh no!</p>
      <p>We need to fix it, but this is an example.</p>
      <p>Your developers and managers should change this page :)</p>
    </>
  ),
}
