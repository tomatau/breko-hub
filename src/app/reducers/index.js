import { combineReducers } from 'redux'
import { routerReducer } from 'react-router-redux'
import { flashReducers as flash } from './flash.reducers'
import { barReducers as bar } from './bar.reducers'

export default combineReducers({
  flash,
  bar,
  routing: routerReducer,
})
