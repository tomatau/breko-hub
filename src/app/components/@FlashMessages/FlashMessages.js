import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { Bem, cleanProps } from 'app/utils'
import { getMessages } from 'app/modules/flash/flash.selectors'
import { removeMessage } from 'app/modules/flash/flash.actions'
import Msg from './Msg'

const bem = new Bem('FlashMessages')

@connect(state => ({
  messages: getMessages(state),
}), { onClickClose: removeMessage })
export default class FlashMessages extends React.Component {
  static propTypes = {
    messages: PropTypes.array,
  };

  static defaultProps = {
    messages: [],
  };

  render() {
    const { messages, onClickClose, ...props } = this.props
    return (
      <div
        {...bem()}
        {...cleanProps(props)}
      >
        {messages.map(msg => (
          <Msg
            key={msg.id}
            msg={msg}
            onClickClose={onClickClose}
          />
        ))}
      </div>
    )
  }
}
