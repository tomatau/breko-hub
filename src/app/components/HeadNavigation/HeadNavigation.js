import React from 'react'
import cx from 'classnames'
import { NavLink, withRouter } from 'react-router-dom'
import { nav as navCopy } from 'app/copy'
import { cleanProps } from 'app/utils'
import styles from './HeadNavigation.module.scss'

export default withRouter(class HeadNavigation extends React.Component {
  render() {
    const { className, ...props } = this.props
    return (
      <nav
        className={cx('HeadNavigation', styles.nav, className)}
        {...cleanProps(props)}
      >
        <NavLink
          exact
          to='/'
          activeClassName={styles.active}
        >
          {navCopy.home}
        </NavLink>
        <NavLink to='/bar' activeClassName={styles.active}>
          {navCopy.bar}
        </NavLink>
        <NavLink to='/private' activeClassName={styles.active}>
          {navCopy.private}
        </NavLink>
      </nav>
    )
  }
})
