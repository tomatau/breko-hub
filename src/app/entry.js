import 'unfetch/polyfill'
import 'raf/polyfill'
import 'react-hot-loader'
import { ConfigService } from 'app/utils'

debug.enable(process.env.DEBUG)

const log = debug('entry')

ConfigService.setEnv(window.__CONFIG_ENV__)
ConfigService.assignVars(window.__APP_CONFIG__)

log('Environment', {
  ...process.env,
  CONFIG_ENV: ConfigService.getEnv(),
  APP_CONFIG: ConfigService.getVars(),
})

require('app/start')
