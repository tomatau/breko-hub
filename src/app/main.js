import { createBrowserHistory, createMemoryHistory } from 'history'
import { ConnectedRouter, routerMiddleware } from 'react-router-redux'
import { Provider } from 'react-redux'
import { inClientViaSocketIO } from 'redux-via-socket.io'

import { middleware, sagaMiddleware } from 'app/composition/middleware'
import { makeCreateStore } from 'app/composition/makeCreateStore'
import socket from 'app/composition/socket'

import rootSaga from 'app/sagas'
import { isBrowser } from 'app/utils'
import rootReducer from 'app/reducers'
import * as app from 'app'

// client store and history
// can go out of sync with server store and server history...
// avoid using directly!
export const history = isBrowser ? createBrowserHistory() : createMemoryHistory()

export const store = makeCreateStore(
  [ ...middleware, routerMiddleware(history) ]
)(rootReducer, isBrowser ? window.__INITIAL_STATE__ : {})

inClientViaSocketIO(socket, store.dispatch)

sagaMiddleware.run(rootSaga)

export const Main = (
  <Provider store={store}>
    <ConnectedRouter history={history}>
      {app.createAppInstance()}
    </ConnectedRouter>
  </Provider>
)

/* istanbul ignore if  */
if (module.hot) {
  /* istanbul ignore next */
  module.hot.accept('app/reducers', () => {
    store.replaceReducer(require('app/reducers'))
  })
}
