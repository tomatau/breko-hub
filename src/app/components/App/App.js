import React from 'react'
import {Â Helmet as DocumentMeta } from 'react-helmet-async'
import { connect } from 'react-redux'
import { Switch, Route, withRouter } from 'react-router-dom'
import { hot } from 'react-hot-loader/root'
import { get } from 'app/utils'
import { addMessage } from 'app/modules/flash/flash.actions'
import { app as appCopy } from 'app/copy'
import { routesList } from 'app/routes'
import HeadNavigation from 'app/components/HeadNavigation/HeadNavigation'
import FlashMessages from 'app/components/@FlashMessages/FlashMessages'
import style from './App.module.scss'

const getLocationFlashState = get('state.flash', {})

@connect(null, { addMessage })
@withRouter
class App extends React.Component {
  componentDidMount() {
    this.addFlashFromLocation()
  }

  componentDidUpdate() {
    this.addFlashFromLocation()
  }

  addFlashFromLocation = () => {
    const { location, addMessage } = this.props
    const flashState = getLocationFlashState(location)
    if (flashState.message) {
      addMessage(flashState.message, flashState.type)
    }
  }

  render() {
    return (
      <>
        <DocumentMeta
          defaultTitle={`${appCopy.title}`}
          titleTemplate={`%s | ${appCopy.title}`}
        >
          <html lang='en' />
          <meta charSet='utf-8' />
          <meta name='description' content={appCopy.meta.description} />
          <meta name='keywords' content={appCopy.meta.keywords} />
          <meta
            name='viewport'
            content='width=device-width,initial-scale=1.0'
          />
        </DocumentMeta>
        <div className={style.app}>
          <HeadNavigation />
          <header>
            <h1>{appCopy.title}</h1>
          </header>
          <main
            aria-live='polite'
            className={style.content}
          >
            <FlashMessages />
            <Switch>
              {routesList.map(route => (
                <Route
                  key={route.name}
                  exact={route.exact}
                  path={route.path}
                  {...route.render
                    ? { render: route.render }
                    : { component: route.component }
                  }
                />
              ))}
            </Switch>
          </main>
        </div>
      </>
    )
  }
}

export default hot(App)
