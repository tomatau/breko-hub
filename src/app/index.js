import App from 'app/components/App/App'

export const createAppInstance = () => <App />
