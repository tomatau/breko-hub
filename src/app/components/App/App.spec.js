import React from 'react'
import { Helmet as DocumentMeta } from 'react-helmet-async'
import { app as appCopy } from 'app/copy'
import HeadNavigation from 'app/components/HeadNavigation/HeadNavigation'
import FlashMessages from 'app/components/@FlashMessages/FlashMessages'
import App from './App'
import styles from './App.module.scss'

describe(`App Component`, function () {
  helpers.setupSnapshots(__filename)

  beforeEach(() => {
    this.wrapper = shallow(<App.WrappedComponent />)
  })

  it(`renders a Helmet document meta as firstChild`, () => {
    const firstChild = this.wrapper.childAt(0)
    expect(firstChild).to.have.type(DocumentMeta)
  })

  it(`renders a div tag with className as second child`, () => {
    const secondChild = this.wrapper.childAt(1)
    expect(secondChild.type()).to.eql('div')
    expect(secondChild).to.have.className(styles.app)
  })

  it(`renders HeadNavigation as first child of div`, () => {
    const firstChild = this.wrapper.childAt(1).childAt(0)
    expect(firstChild).to.have.type(HeadNavigation)
  })

  it(`renders a header as second child of div`, () => {
    expect(
      snap(this.wrapper.childAt(1).childAt(1))
    ).to.matchSnapshot()
  })

  it(`renders a main.content[aria-live] as third child of div`, () => {
    const thirdChild = this.wrapper.childAt(1).childAt(2)
    expect(thirdChild).to.have.type('main')
    expect(thirdChild).to.have.className(styles.content)
    expect(thirdChild).to.have.prop('aria-live', 'polite')
  })

  describe(`DocumentMeta`, () => {
    it(`sets the meta options`, () => {
      expect(this.wrapper.find(DocumentMeta).getElement()).to.eql(
        <DocumentMeta
          defaultTitle={`${appCopy.title}`}
          titleTemplate={`%s | ${appCopy.title}`}
        >
          <html lang='en' />
          <meta charSet='utf-8' />
          <meta name='description' content={appCopy.meta.description} />
          <meta name='keywords' content={appCopy.meta.keywords} />
          <meta name='viewport' content='width=device-width,initial-scale=1.0' />
        </DocumentMeta>
      )
    })
  })

  describe(`Title`, () => {
    it(`should have app title as text`, () => {
      const title = this.wrapper.find('h1')
      expect(title.text()).to.contain(appCopy.title)
    })
  })

  describe(`Main`, () => {
    it(`renders FlashMessages as first child of main`, () => {
      const firstChild = this.wrapper.find('main').childAt(0)
      expect(firstChild).to.have.type(FlashMessages)
    })

    it(`renders routes inside Main`, () => {
      expect(
        snap(this.wrapper.find('main'))
      ).to.matchSnapshot()
    })
  })
})
