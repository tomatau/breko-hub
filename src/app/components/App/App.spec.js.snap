// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Main renders routes inside Main 1`] = `
<main
  aria-live="polite"
  className="src-app-components-App-App-module-content"
>
  <Connect(FlashMessages) />
  <Switch>
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "load": [Function],
          "preload": [Function],
          "render": [Function],
        }
      }
      exact={true}
      key="home"
      path="/"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "load": [Function],
          "preload": [Function],
          "render": [Function],
        }
      }
      key="bar"
      path="/bar"
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "load": [Function],
          "preload": [Function],
          "render": [Function],
        }
      }
      key="oops"
      path="/oops"
    />
    <Route
      key="private"
      path="/private"
      render={[Function]}
    />
    <Route
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "load": [Function],
          "preload": [Function],
          "render": [Function],
        }
      }
      key="notFound"
    />
  </Switch>
</main>
`;

exports[`App Component renders a header as second child of div 1`] = `
<header>
  <h1>
    Breko Hub
  </h1>
</header>
`;
